### PR TITLE
KAFKA-3849: Add explanation on why polling every second in MirrorMaker is required.

### PR DESCRIPTION
**Pull request by SinghAsDev**

### CODE REVIEW

File: core/src/main/scala/kafka/tools/MirrorMaker.scala [hunk 1]
```diff
@@ -184,6 +191,10 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
           System.exit(1)
         }
       } else {
+        if (options.has(pollMsOpt)) {
+          error(s"$pollMsOpt can only be used with $useNewConsumerOpt.")
```
Comment (granthenke) [inline core/src/main/scala/kafka/tools/MirrorMaker.scala]: The jira states the old consumer supports this. Can we pass this value to the old consumer too?

**Comment by granthenke** [comment]
I am not super convinced on how valuable of an option this is, or why someone would use it. But I guess the option can't hurt.

Did you test MM after these changes? Based on visual inspection I had one small comment, otherwise it LGTM.

**Comment by hachikuji** [comment]
Yeah, I'm also a little puzzled why this is needed.

**Comment by SinghAsDev** [comment]
Hey Grant/ Jason, thanks for the quick review. If you have a pipeline where data arrives say every few minutes, polling every second seems to be extraneous. Having that said, increasing poll time is not going to save lots of CPU cycles, so it is up for debate if this is worth a new config or not. In old consumer one would have controlled this using `consumer.timeout.ms` and adding this config will allow users to set the poll time for new consumers as well.

**Comment by hachikuji** [comment]
@SinghAsDev Hey Ashish, to be honest, I'm not sure why we're using 1 second as the poll timeout. Would there be any downside to just increasing it? From a glance at the code, it seems like we catch the exception and retry anyway. Also, one iteration of the loop per second seems like it would have minimal impact on CPU.

**Comment by SinghAsDev** [comment]
@hachikuji increasing timeout seems to be reasonable. If someone is getting high speed data then they do not have worry about the timeout anyway. However, that again raises the question what is an optimal value for timeout, IMO it depends on incoming data speed. What do you think is a good default value?

**Comment by hachikuji** [comment]
I would probably use `Long.MAX_VALUE`. We break from poll() as soon as any data is available, so there's no obvious downside to blocking in poll() when there's nothing available. Unless maybe we depend on the timeouts to trigger some periodic behavior?

**Comment by SinghAsDev** [comment]
I do not see any behavior depending on `ConsumerTimeoutException` in MirrorMaker, it just gets trace logged. @becketqin would have any suggestion on this?

**Comment by SinghAsDev** [comment]
@hachikuji updated PR to instead increase poll time. Sanity tested the change.

**Comment by hachikuji** [comment]
@SinghAsDev LGTM. Would be nice to hear @becketqin's thoughts to make sure there weren't any good reasons to keep the timeout low.

**Comment by granthenke** [comment]
I haven't seen a case or report why the current value could be or is problematic. What reason do we have to change it? Is there zero risk in changing it? 

If my understanding is correct, increasing the timeout above the request timeout or metadata timeout just means that the looping will now move from MM to the [while loop inside of the Consumer](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L936) because of [how the Network client poll works](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/NetworkClient.java#L257). 

Because request.timeout.ms defaults to 30 seconds and metadata.max.age.ms defaults to 5 minutes. I think the longest we would be saving a CPU cycle is 30 seconds regardless of the poll timeout. I could be interpreting the code wrong though.

**Comment by SinghAsDev** [comment]
@granthenke this came up as a related question from customer. I think what we are comparing here is, poll every second even if no data is arriving, vs return when sufficient amount of data has arrived (which is configurable). Sure, it is not going to save significant amount of CPU and that is stated in previous comments. I agree that this change is not improving behavior of MM by a lot, and so if there is some risk associated with it, we should drop the change. However, I will be interested to know what risk you see with the suggested change.

**Comment by hachikuji** [comment]
Yeah, I'm a little doubtful of any real CPU impact. As Grant mentions, the consumer hides its own loop behind the poll API, and the frequency is bound basically by the heartbeat interval. That said, I tend to see polling with a large timeout as a better pattern, though this is largely my own preference. It would let us drop the hasNext() check on the following line, but it seems not too big of a deal either way.

**Comment by SinghAsDev** [comment]
I can be convinced either ways on this. Let's wait on @becketqin feedback as well, and we can close it accordingly.

**Comment by xiaotao183** [comment]
Increasing the poll time to `Long.MAX_VALUE` would have impact on offset commit especially for low volume topic. Any uncommitted messages before entering `poll(Long.MAX_VALUE)` will not be committed until a new message arrives.  We actually breaks the contract of `offset.commit.interval.ms`

**Comment by hachikuji** [comment]
@xiaotao183 Great point. It seems we could either leave this as it is or we could use `offset.commit.interval.ms` as the timeout passed to `poll()`. It's probably also worth adding a comment in the code explaining the dependence so that we're not tempted to make the same change again in the future.

**Comment by SinghAsDev** [comment]
@xiaotao183 thanks for pointing it out. I agree with @hachikuji that we should add a comment in the code. Also, I am inclined towards using `offset.commit.interval.ms` as `poll()` timeout, which by default is a minute. Makes poll time configurable without adding a new config and does not have an infinite timeout. Does it sound reasonable to you @granthenke @hachikuji @xiaotao183 @becketqin ?

**Comment by hachikuji** [comment]
@SinghAsDev Sounds good to me.

**Comment by granthenke** [comment]
I won't get in the way of this change going in, but I am still unconvinced. 

Through extended discussion, we found that there would have been a negative impact for a change with near 0 positive impact (that I know of). Who is to say there is not another thing we are not seeing or thinking of? Since there is no real benefit what is the value of this change and risk?

I would love to see more Mirrormaker testing so that we can be confident in our changes. But as its stands today we have 2 ducktape tests (test_simple_end_to_end, test_bounce).

**Comment by hachikuji** [comment]
Actually I think changing the poll timeout to `offset.commit.interval.ms` is justifiable because of the same problem mentioned by @xiaotao183. If the user specifies 500ms, for example, then we won't actually be able to enforce it unless the poll timeout is any higher than that. Now, sub-second commit intervals are pretty unlikely in practice (I guess), but it still seems more appropriate to use the commit interval rather than a somewhat arbitrary hard-coded value. At the very least, we should add the comment to avoid making the same mistake again in the future.

**Comment by SinghAsDev** [comment]
The failed test, `KStreamRepartitionJoinTest.shouldCorrectlyRepartitionOnJoinOperations` seems unrelated. It passes locally. Did not find a JIRA for marking this as an unstable test, however I could not reproduce test failure locally, let me know if it still warrants a unstable test JIRA. @ijuma would you be able to review this one?

**Comment by ijuma** [comment]
Yes, a JIRA for the unstable test would be appreciated. With regards to this change, I share some of Grant's concerns given the lack of existing MirrorMaker tests and the fact that the benefit seems a bit esoteric. Also, no test is included in this PR.

Any thoughts @becketqin?

**Comment by SinghAsDev** [comment]
Filed KAFKA-3896 to track the unstable test.

The original issue, which I agree was not really causing any obvious
problem, led to uncovering of offset commits happening faster than
specified `offset.commit.interval.ms`. Though that is again not an obvious
issue, but certainly an unexpected behavior. Review from @becketqin will be
helpful. Also, on the tests side, I am trying to think how can this be
tested. Would appreciate any suggestion on that.

**Comment by hachikuji** [comment]
Would have been better if the JIRA spelled out an actual problem. I'm OK leaving it as it is or using the offset commit interval for the poll timeout (which seems intuitive and low-risk). At the very least, it would be nice to have the comments which clarify why the poll interval must be small since it is not obvious just looking over the code.

**Comment by ijuma** [comment]
@SinghAsDev Offset commits don't happen faster than specified as far as I can see. The potential issue is if you set the `offset.commit.interval.ms` to be _lower_ than 1 second (the hardcoded poll timeout), then the offset commit may not happen within the time you specify if poll ends up waiting for 1 second because no messages have been received (for example). Of course, this only matters if there are uncommitted messages since the previous poll, which again needs a specific sequence of events (a poll that takes less time than the `offset.commit.interval.ms` so that the messages returned are not immediately committed).

Am I missing something?

**Comment by ijuma** [comment]
@hachikuji, I agree that we should include the comment no matter what.

**Comment by SinghAsDev** [comment]
@ijuma my bad, you are right. If you folks think using `offset.commit.interval.ms` is very risky then I will be fine with just adding an explanation on why the hard coded value.

**Comment by SinghAsDev** [comment]
Updated PR and Issue title to just add explanation on why hard coded value is used. We can take up the issue of using `offset.commit.interval.ms` as poll timeout when and if we see any issue related with it. @ijuma mind giving it another pass.

**Comment by ijuma** [comment]
LGTM, merged to trunk.

**Pull request closed without merging**